### PR TITLE
Clean up, improved test coverage and documentation around TransformedClassPath

### PR DESCRIPTION
**Pull request by mlopatkin**
Part of #23297

### CODE REVIEW

**Comment by mlopatkin** [comment]
@bot-gradle test this

File: subprojects/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java [hunk 1]
```diff
@@ -31,11 +31,16 @@
...
+/**
+ * A special ClassPath that contains both original JARs/class directories and their bytecode-transformed counterparts.
+ * The class loaders constructed from this classpath can replace classes with transformed ones when loading.
+ */
 public class TransformedClassPath implements ClassPath {
     private final ClassPath originalClassPath;
     private final ImmutableMap<File, File> transforms;
 
     private TransformedClassPath(ClassPath originalClassPath, Map<File, File> transforms) {
+        assert !(originalClassPath instanceof TransformedClassPath);
```
Comment (bamboo) [inline subprojects/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java:43]: 👍

File: subprojects/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java [hunk 2]
```diff
@@ -65,53 +85,90 @@ public List<File> getAsTransformedFiles() {
...
+     * Prepends the given classPath to this classpath.
+     * No transformations are applied to the prepended class path.
+     * <p>
+     * This is a helper to implement DefaultClassPath.plus(TransformedClassPath) efficiently.
+     *
+     * @param classPath the classPath to prepend to this classpath
+     * @return the new transformed classpath
+     */
+    TransformedClassPath prepend(DefaultClassPath classPath) {
+        // If some entries from this classpath are also in the prepended classpath, then the transform should be undone in the result.
```
Comment (bamboo) [inline subprojects/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java:118]: ❓ I don't get _then the transform should be undone in the result_, could you rephrase or expand?

File: subprojects/base-services/src/test/groovy/org/gradle/internal/classpath/TransformedClassPathTest.groovy [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+        then:
+        combined.asFiles == [file("1.jar"), file("2.jar")]
+        (combined as TransformedClassPath).asTransformedFiles == [file("1.jar"), file("t2.jar")]
+
+        combined.findTransformedJarFor(file("1.jar")) == null
+        combined.findTransformedJarFor(file("2.jar")) == file("t2.jar")
+    }
+
+    def "non-transformed jar on transformed classpath stays non-transformed when another transformation is appended"() {
```
Comment (bamboo) [inline subprojects/base-services/src/test/groovy/org/gradle/internal/classpath/TransformedClassPathTest.groovy:126]: ❓ What's the rationale for this rule? Intuitively I'd expect a transformed jar to override a non-transformed jar.

**Review by bamboo — CHANGES_REQUESTED** [review]

File: subprojects/base-services/src/test/groovy/org/gradle/internal/classpath/TransformedClassPathTest.groovy (hunk 3 above)
Comment (mlopatkin) [reply subprojects/base-services/src/test/groovy/org/gradle/internal/classpath/TransformedClassPathTest.groovy:126]: It follows the behavior that you get in the class loader when you combine transformed and non-transformed JARs with concatenation (or as `DefaultClassPath` instances). AFAIK, the jars in the classpath are searched left-to-right, so the first jar with the matching class wins. If we were to combine `DefaultClassPath` instances, we would get:
```
nonTransformed = DCP(1.jar)
transformed1 = DCP(t2.jar)
transformed2 = DCP(t1.jar)

combined = (nonTransformed + transformed1) + transformed2 => 1.jar:t2.jar:t1.jar
```
As `1.jar` and `t1.jar` contain the same classes, each class loading would end up loading a class from `1.jar`, as it comes first in the combined class path. To keep this behavior, `TransformedClassPath` has to drop the transformation for `1.jar` that comes from `transformed2`.

File: subprojects/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java (hunk 2 above)
Comment (mlopatkin) [reply subprojects/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java]: I've tried to give a more detailed explanation in the class javadoc. It should also reduce the confusion in the test case https://github.com/gradle/gradle/pull/23351#discussion_r1086626689, as these are actually closely related.

File: subprojects/base-services/src/main/java/org/gradle/internal/classpath/ClassPath.java [hunk 4]
```diff
@@ -31,19 +31,63 @@ public interface ClassPath {
 
     ClassPath EMPTY = new DefaultClassPath();
 
+    /**
+     * Returns {@code true} iff the class path is empty.
```
[the remark this answers is not in the document]
Comment (mlopatkin) [reply subprojects/base-services/src/main/java/org/gradle/internal/classpath/ClassPath.java]: Thanks, I rephrased it a bit.

File: subprojects/base-services/src/main/java/org/gradle/internal/agents/InstrumentingClassLoader.java [hunk 5]
```diff
@@ -19,9 +19,36 @@
...
+     * The {@code classfileBuffer} may be already modified by other Java agents.
+     *
+     * @param className the name of the class being loaded (in the internal form, e.g. {@code java/util/List})
+     * @param protectionDomain the protection domain of the original class, as defined by this classloader
+     * @param classfileBuffer the buffer that contains class implementation bytes in class file format - must not be modified
+     * @return new class implementation bytes in class file format or {@code null} to continue loading the original implementation
+     */
     @Nullable
-    byte[] instrumentClass(@Nullable String className, ProtectionDomain protectionDomain, byte[] classfileBuffer);
+    byte[] instrumentClass(@Nullable String className, @Nullable ProtectionDomain protectionDomain, byte[] classfileBuffer);
```
[the remark this answers is not in the document]
Comment (mlopatkin) [reply subprojects/base-services/src/main/java/org/gradle/internal/agents/InstrumentingClassLoader.java:48]: > What about thread-safety? Can this method be called from multiple threads? Can this method be called while `transformFailed` is being called for another request?

I added some wording about concurrency to the javadocs. In short, yes, these methods may be called concurrently.

File: subprojects/base-services/src/test/groovy/org/gradle/internal/classpath/TransformedClassPathTest.groovy (hunk 3 above)
Comment (bamboo) [reply subprojects/base-services/src/test/groovy/org/gradle/internal/classpath/TransformedClassPathTest.groovy:126]: Why not move `t1.jar` to be the transformed counterpart of the early `1.jar` and serve the transformed classes from that earlier point in the classpath?

File: subprojects/base-services/src/main/java/org/gradle/internal/agents/InstrumentingClassLoader.java [hunk 6]
```diff
@@ -19,9 +19,42 @@
 import javax.annotation.Nullable;
 import java.security.ProtectionDomain;
 
+/**
+ * A ClassLoader implementing this interface gets a special treatment from the instrumenting Java agent.
+ * When such a classloader attempts to define a class, the agent calls {@link #instrumentClass(String, ProtectionDomain, byte[])} method
+ * giving the classloader a chance to rewrite the bytecode of the class.
+ * <p>
+ * Methods defined in this interface may be called concurrently in multiple threads.
```
Comment (bamboo) [inline subprojects/base-services/src/main/java/org/gradle/internal/agents/InstrumentingClassLoader.java:27]: 👍

File: subprojects/base-services/src/main/java/org/gradle/internal/agents/InstrumentingClassLoader.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bamboo) [reply subprojects/base-services/src/main/java/org/gradle/internal/agents/InstrumentingClassLoader.java:48]: 👍

**Review by bamboo — APPROVED** [review]

**Comment by mlopatkin** [comment]
@bot-gradle test and merge

**Comment by mlopatkin** [comment]
@bot-gradle test QF

**Comment by mlopatkin** [comment]
@bot-gradle test and merge
